tour_model_eval/evaluation_pipeline.py

```diff
@@ -0,0 +1,173 @@
+    percentage_second = grp.get_req_pct(new_labels, track, filter_trips, sim)
+
+    # get homogeneity score for the second round
+    homo_second = gs.score(bin_trips, new_labels)
+    return percentage_second,homo_second
+
+
+# we use functions in similarity to build the first round of clustering
+def first_round(data,radius):
+    sim = similarity.similarity(data, radius)
+    filter_trips = sim.data
+    sim.bin_data()
+    sim.delete_bins()
+    bins = sim.bins
+    bin_trips = sim.newdata
+    return sim, bins, bin_trips, filter_trips
+
+
+def get_first_label(bins):
+    # get first round labels
+    # the labels from the first round are the indices of bins
+    # e.g. in bin 0 [trip1, trip2, trip3], the labels of this bin is [0,0,0]
+    first_labels = []
+    for b in range(len(bins)):
+        for trip in bins[b]:
+            first_labels.append(b)
+    return first_labels
+
+
+def get_track(bins, first_labels):
+    # create a list idx_labels_track to store indices and labels
+    # the indices of the items will be the same in the new label list after the second round clustering
+    # item[0] is the original index of the trip in filter_trips
+    # item[1] is the label from the first round of clustering
+    idx_labels_track = []
+    for bin in bins:
+        for ori_idx in bin:
+            idx_labels_track.append([ori_idx])
+    # store first round labels in idx_labels_track list
+    for i in range(len(first_labels)):
+        idx_labels_track[i].append(first_labels[i])
+
+    return idx_labels_track
+
+
+def tune(data,radius):
+    sim, bins, bin_trips, filter_trips = first_round(data, radius)
+    # it is possible that we don't have common trips for tuning or testing
+    # bins contain common trips indices
+    if len(bins) is not 0:
+        gs.compare_trip_orders(bins, bin_trips, filter_trips)
+        first_labels = get_first_label(bins)
+        # new_labels temporary stores the labels from the first round, but later the labels in new_labels will be
+        # updated with the labels after two rounds of clustering.
+        new_labels = first_labels.copy()
+        first_label_set = list(set(first_labels))
+        track = get_track(bins, first_labels)
+        # collect tuning scores and parameters
+        tune_score = {}
+        for dist_pct in np.arange(0.15, 0.6, 0.02):
+            for low in range(250, 600):
+                percentage_second, homo_second = second_round(first_label_set, first_labels, bin_trips, filter_trips,
+                                                              low, dist_pct,
+                                                              sim, new_labels, track)
+
+                curr_score = gs.get_score(homo_second, percentage_second)
+                if curr_score not in tune_score:
+                    tune_score[curr_score] = (low, dist_pct, homo_second, percentage_second)
+
+        best_score = max(tune_score)
+        sel_tradeoffs = tune_score[best_score][0:2]
```

Comment:
again, it seems like it would be good to create a class called ClusteringResults with these fields or even return a simple dict with the fields instead of passing around lists that are hard to read. Lists are convenient for you to write, but code like `[0:2]` doesn't indicate to others what you are returning.

Reply:
I have already changed this part, there is no more`[0:2]`. I do have a dict to collect the parameters. The parameters are clearly pointed out and return separately. 
e.g.     return low,dist_pct